React Revamp: What is JSX lesson (#24652)
TheOdinProject/curriculum: react/getting_started_with_react/what_is_jsx/what_is_jsx.md

   Correct:

   ~~~jsx
   function App() {
     return (
       <>
         <div className="container">
           <svg>
             <circle cx="25" cy="75" r="20" stroke="green" strokeWidth="2" />
           </svg>
         </div>
       </>
     );
   }
   ~~~

   Incorrect:

   ~~~jsx
   function App() {
     return (
       <>
         <div class="container">
           <svg>
             <circle cx="25" cy="75" r="20" stroke="green" stroke-width="2" />
           </svg>
         </div>
       </>
     );
   }
   ~~~

### Converting HTML to JSX

Now that we've covered the Rules of JSX, we'll go through the conversion of a chunk of HTML to JSX.

~~~jsx
<h1>Test title</h1>
<ol class="test-list">
  <li>List item 1
  <li>List item 2
  <li>List item 3
</ol>
<svg >
   <circle cx="25" cy="75" r="20" stroke="green" stroke-width="2" />
</svg>
<form><input type="text"></form>
~~~

> If you try to return this from a React component, you would get many errors, so we are going to fix that! We would recommend following along with this in your own React
> environment.(If you don't want to create a new React app locally, go to [react.new](https://react.new/) to have a quick React environment up and going in your web browser.

Reviewer: Learners should have a React project setup locally from the previous lessons. Although I do like the exposing them to react.new

Wording suggestion to accommodate the above:
```suggestion
If you try to return this from a React component, you would get many errors, so we are going to fix that! We would recommend following along with this in your own React environment.

 Make sure to follow along this example in your local environment. Alternatively, you can go to [react.new](https://react.new/) to have a quick React environment up and going in your web browser.
```

Author: Seems like a sound change, will commit the suggestion 👍🏻

### Introduction

You're likely somewhat familiar with JSX from previous React lessons, in this lesson we're going to explain _what_ it is, _why_ we use it, and _how_ to use it in your React apps.

### Lesson Overview

This section contains a general overview of topics that you will learn in this lesson.

- What JSX is.
- Differences between JSX and HTML.

### What is JSX?

JSX is a syntax extension for Javascript that lets you write HTML-like markup inside a Javascript file. It's not required to use JSX when writing React components, but it does make writing them more simple and concise.

Essentially, JSX is syntactic sugar for the React [createElement](https://reactjs.org/docs/react-api.html#createelement) function. This function creates a React element, which is a plain object, so JSX compiles down to plain JavaScript objects.

Here you can see the logged value of a div JSX element with some text:

> ![JSX element's logged value](./imgs/00.png)

Reviewer: Just leaving this comment here so we don't forget.

After the lesson is merged, this image link needs to be replaced with a statically link. Our guide on [adding images in the curriculum](https://github.com/TheOdinProject/curriculum/wiki/Adding-Images-to-the-Curriculum) can be followed.

### Introduction

You're likely somewhat familiar with JSX from previous React lessons, in this lesson we're going to explain _what_ it is, _why_ we use it, and _how_ to use it in your React apps.

### Lesson Overview

This section contains a general overview of topics that you will learn in this lesson.

> - What JSX is.
> - Why we use JSX.
> - Differences between JSX and HTML.
> - How to write and take advantage of JSX's features.

Reviewer: We can clean this up:

```suggestion
- What JSX is.
- Differences between JSX and HTML.
```

### What is JSX?

JSX is a syntax extension for Javascript that lets you write HTML-like markup inside a Javascript file. It's not required to use JSX when writing React components, but it does make writing them more simple and concise.

Essentially, JSX is syntactic sugar for the React [createElement](https://reactjs.org/docs/react-api.html#createelement) function. This function creates a React element, which is a plain object, so JSX compiles down to plain JavaScript objects.

Here you can see the logged value of a div JSX element with some text:

![JSX element's logged value](./imgs/00.png)

### Why do we use JSX?

Most of the time in apps, rendering logic and markup are inherently coupled, but we typically separate them by having our logic and markup in separate files. JSX allows React to separate concerns by containing both rendering logic and content in the same place (a component).

On top of this, it's a more intuitive, visual way to work with the UI inside your code, and allows React to show more useful error and warning messages.

> ### Converting HTML to JSX
> 
> If you were to take some valid HTML and copy it straight into your React component, it would not work. This is due to some of the rules JSX implements, that aren't present in HTML.
> 
> #### Rules of JSX:

Reviewer: I think we should make Rules of JSX a first level heading, and move the "Converting HTML to JSX" heading after the rules section.

```suggestion
### Rules of JSX

If you were to take some valid HTML and copy it straight into your React component, it would not work. This is due to some of the rules JSX implements, that aren't present in HTML.
```

1. Return a single root element.

   If you wish to return multiple elements in a component, you can do so by wrapping them in a parent tag. This can be a `<div>`, or, if you don't want the elements to have a container, you could use a [React fragment](https://beta.reactjs.org/apis/react/Fragment), like so: `<>Children</>`

   Correct:

   ~~~jsx
   function App() {
     return (
       <>
         <h1>Example h1</h1>
         <h2>Example h2</h2>
       </>

>        // Could replace <></>with <div></div>f

Reviewer: ```suggestion
       // Could replace <></>with <div></div>
```

   Correct:

   ~~~jsx
   function App() {
     return (
       <>
         <input />
         <li></li>
       </>
     );
   }
   ~~~

   Incorrect:

   ~~~jsx
   function App() {
     return (
         <>
          <input>
          <li>
         </>
     );
   }
   ~~~

3. camelCase **Most** things.

   JSX turns into JavaScript, and attributes of elements become keys of JavaScript objects, so you can't use dashes or reserved words such as `class`. Because of this, many HTML attributes are written in camelCase. Instead of `stroke-width`, you'd use `strokeWidth`, and instead of `class` you'd use `className`.

   Correct:

   ~~~jsx
   function App() {
     return (
       <>
         <div className="container">
           <svg>
             <circle cx="25" cy="75" r="20" stroke="green" strokeWidth="2" />
           </svg>
         </div>
       </>
     );
   }
   ~~~

   Incorrect:

   ~~~jsx
   function App() {
     return (
       <>
         <div class="container">
           <svg>
             <circle cx="25" cy="75" r="20" stroke="green" stroke-width="2" />
           </svg>
         </div>
       </>
     );
   }
   ~~~

> Now that we've covered the Rules of JSX, we'll go through the conversion of a chunk of HTML to JSX.

Reviewer: ```suggestion
### Converting HTML to JSX

Now that we've covered the Rules of JSX, we'll go through the conversion of a chunk of HTML to JSX.
```
